ethyca/fides
# Populate Application DB instead of the Test DB

❗ This could create breaking changes in the -plus repos that would need to be handled before they are upgraded to use this 

Closes #1175 

### Code Changes

Mostly ctl-side changes: 

#### Environment Variables

- Don't set `FIDES__TEST_MODE` in the `docker-compose.yml` - this should be **set by pytest only**. So this is _only True when running pytest_. 
- Update incorrect variable name `TESTING` to  `FIDES__TEST_MODE` in `pyproject.toml` to set test mode to true for pytest only.  `TESTING` was the old variable we used in ops.
- Update base credentials > app_postgres to be a connection string pointing to the `fides` db instead of the `fides_test` db in the `.fides/fides.toml`.  Running `fides scan dataset db --credentials-id app_postgres` locally should scan the application db not the test db.


#### Sessions
- Dependency injection: Inject an async session as a dependency into the ctl `crud`, `datamap`, `view`, and `generate` routes.  Pass that session onto the crud methods.  (Ops already injects a session into its routes. ) The admin route doesn't seem to need a session because it is using database_uri's that have test_mode awareness. The `health` endpoint is already injecting the sync session from ops.  The validate routes don't seem to need a session either.
    - Create an `async_db` generator for API endpoint dependency injection.  
- Update the ctl database crud methods to take in an `async_session` as a variable (`api/ctl/database/crud.py`), which is test_mode aware
- Update ctl route utils to take in an `async_session` parameter (`api/ctl/routes/util.py`) that is test mode aware


#### Test refactors

- Add an async_session to the ctl conftest for use in testing methods that need an async_session
- Remove unused user fixtures from the ctl conftest
- Monkeypatch several methods in the `requests` library to use the `TestClient` (which calls functions directly instead of sending a request to a running webserver). This is primarily for the methods in `src/fides/ctl/core/api.py`.  Now all the CLI-related tests and all methods calling `api.get `or `api.pos`t etc. should be usingthe TestClient instead of the requests library.


#### Ops changes
Ops was already using a `TestClient` to test its endpoints, which does not ping a running webserver.

- Update ops test  `setup_db` to reset the test db, not the application db (by using the test client).


#### Misc 
- Fix existing logging tests to remove need for unnecessary `toggle_testing_envvar` fixture
- Have `logging_sessions` log debug logs on dev mode, not test mode (test mode is false now, when running application locally). Running the application had been resulting in fewer logs now that test mode was false.
- Have `load_default_organization` and `load_default_taxonomy` take in an async_session (ctl-side)



### Steps to Confirm

* [x] Run `nox -s dev -- postgres` to bring up the webserver
* [x] Connect to the `fides` application db `psql -h localhost -p 5432 -U postgres -d fides` locally and verify that this application db is up and populated
* [x] Run a privacy request against the postgres db and verify the supporting resources are all created in the application db
* [x] Try to connect to the test db and observe it doesn't exist yet `psql -h localhost -p 5432 -U postgres -d fides_test`
* [x] Now bring up the shell `nox -s dev -- postgres shell`
* [x] Run some tests in pytest `pytest tests/ctl/database/test_crud.py`, put a breakpoint in the test after resources are created
* [x] Connect to the test db ``psql -h localhost -p 5432 -U postgres -d fides_test`.  Observe the test db was created, and the resources exist in the test db
* [x] Query the application db - observe that it is not wiped, and previously created resources still exist
* [x] Run `fides get data_category user` - observe successful return from application db

### Pre-Merge Checklist

* Documentation Updated:
  * [ ] documentation complete, or draft/outline provided (tag docs-team to complete/review on this branch)
  * [ ] documentation issue created (tag docs-team to complete issue separately)

### Description Of Changes

Fides is currently writing to the test db both when running the application and when we're running pytest.  This is difficult for local development - when you run tests in pytest, the test db is wiped and populated with fixture data.  It would be useful to populate the application db when running the application locally and run the test db when running pytest. 

Note that before the merge of fidesctl and fidesops, fidesops already had this division, writing to the application db or the test db depending on the context .  Fidesctl differed in that it spun up a webserver and the tests made requests against that running webserver. Fidesops used test clients.  This refactor primarily injects sessions into ctl-routes, and then uses test clients that have awareness of test mode to talk to the test db during testing.  However, for the CLI tests that are talking to a running webserver, we're monkeypatching the `requests` library to use the TestClient instead.


### Note

I want to make sure we're not being too aggressive with the monkey patching - ops uses the requests library to make connector requests in many places. I am only monkeypatching the requests library ctl side.
